internal/primitives/runtime/generic/header.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package generic
+
+import (
+	"io"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Header is a block header, and implements a compatible encoding to `sp_runtime::generic::Header`
+type Header[N runtime.Number, H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	// The parent hash.
+	parentHash H
+	// The block number.
+	number N
+	// The state trie merkle root
+	stateRoot H
+	// The merkle root of the extrinsics.
+	extrinsicsRoot H
+	// A chain-specific digest of data useful for light clients or referencing auxiliary data.
+	digest runtime.Digest
+}
+
+// Number returns the block number.
+func (h Header[N, H, Hasher]) Number() N {
+	return h.number
+}
+
+// SetNumber sets the block number.
+func (h *Header[N, H, Hasher]) SetNumber(number N) {
+	h.number = number
+}
+
+// ExtrinsicsRoot returns the extrinsics root.
+func (h Header[N, H, Hasher]) ExtrinsicsRoot() H {
+	return h.extrinsicsRoot
+}
+
+// SetExtrinsicsRoot sets the extrinsics root.
+func (h *Header[N, H, Hasher]) SetExtrinsicsRoot(root H) {
+	h.extrinsicsRoot = root
+}
+
+// StateRoot returns the state root.
+func (h Header[N, H, Hasher]) StateRoot() H {
+	return h.stateRoot
+}
+
+// SetStateRoot sets the state root.
+func (h *Header[N, H, Hasher]) SetStateRoot(root H) {
+	h.stateRoot = root
+}
+
+// ParentHash returns the parent hash.
+func (h Header[N, H, Hasher]) ParentHash() H {
+	return h.parentHash
+}
+
+// SetParentHash sets the parent hash.
+func (h *Header[N, H, Hasher]) SetParentHash(hash H) {
+	h.parentHash = hash
+}
+
+// Digest returns the digest.
+func (h Header[N, H, Hasher]) Digest() runtime.Digest {
+	return h.digest
+}
+
+// DigestMut returns a mutable reference to the stored digest.
+func (h Header[N, H, Hasher]) DigestMut() *runtime.Digest {
+	return &h.digest
+}
+
+type helper[H any] struct {
```

Comment:
A thought I wanted to throw out there: does polkadot actually use multiple Hash types in its implementation?

If not, perhaps we remove generics for Hash and use a concrete type and simplify the code

Reply:
It's actually pretty useful to accommodate any generic hash when testing individual components. See the `LeafSet` tests as an example. It would introduce extra complexity in the tests to ensure everything is type `hash.H256` since that is what is used in practice.

Reply:
This is true, however, this is what we do in gossamer currently and it is not to much additional complexity. So perhaps its worth sacrificing a bit of complexity in tests to allow for less complex production code